_drafts/kotlin-coroutines-101.md

```diff
@@ -0,0 +1,915 @@
+
+On the other end, as we will see, **coroutines are very lightweight**. They are not mapped directly on OS threads but at the user level, with simple objects called _continuations_. Switching between coroutines does not require the OS to load another thread's context but to switch the reference to the continuation object.
+
+Another good reason to adopt **coroutines** is that they **are a way to write asynchronous code in a synchronous fashion**.
+
+In the first attempt, we can use callbacks. However, callbacks are not very elegant, and they are not composable. Moreover, they are not very easy to reason about them. It's easy to end up in a *callback hell*, where the code is tough to read and maintain:
+
+
+```kotlin
+a(aInput) { resultFromA ->
+  b(resultFromA) { resultFromB ->
+    c(resultFromB) { resultFromC ->
+      d(resultFromC) { resultFromD ->
+        println("A, B, C, D: $resultFromA, $resultFromB, $resultFromC, $resultFromD")
+      }
+    }
+  }
+}
+```
+
+The example above shows the execution of four functions using the callback style. As we can see, collecting the four values returned by the four functions takes a lot of work. Moreover, the code could be easier to read and maintain.
+
+Another model that is used in asynchronous programming is reactive programming. However, the problem is that it needs to produce more complex code to understand and maintain. Let's take, for example, the following code snippet from the official documentation of the RxJava library:
+
+```java
+Flowable.fromCallable(() -> {
+    Thread.sleep(1000); //  imitate expensive computation
+    return "Done";
+})
+  .subscribeOn(Schedulers.io())
+  .observeOn(Schedulers.single())
+  .subscribe(System.out::println, Throwable::printStackTrace);
+```
+
+The above code simulates the run of some computation, and network request on a background thread, showing the results (or error) on the UI thread. It's not self-explanatory, and we need to know the library well to understand what's happening.
+
+Coroutines solve all the above problems. Let's see how.
+
+## 3. Suspending Functions
+
+As we said, a coroutine is a lightweight thread, which means it's not mapped directly to an OS thread. It's a computation that can be suspended and resumed at any time. So, before we can start looking at how to build a coroutine, we need to understand how to suspend and resume a coroutine.
+
+**Kotlin provides the `suspend` keyword to mark a function that can suspend a coroutine**:
```

Comment:
"... that can suspend a coroutine, i.e. allow it to be paused & resumed later"
So that we avoid using the word in its own definition